src/store/message-store-level.ts

```diff
@@ -77,19 +82,149 @@ export class MessageStoreLevel implements MessageStore {
     return message;
   }
 
-  async query(tenant: string, filter: Filter, options?: MessageStoreOptions): Promise<GenericMessage[]> {
+  async query(
+    tenant: string,
+    filters: Filter[],
+    messageSort?: MessageSort,
+    pagination?: Pagination,
+    options?: MessageStoreOptions
+  ): Promise<{ messages: GenericMessage[], paginationMessageCid?: string }> {
     options?.signal?.throwIfAborted();
 
     const messages: GenericMessage[] = [];
+    const resultIds = await this.index.query(filters.map(f => ({ ...f, tenant })), options);
 
-    const resultIds = await this.index.query({ ...filter, tenant }, options);
-
+    // as an optimization for large data sets, we are finding the message object which matches the paginationMessageCid here.
+    // we can use this within the pagination function after sorting to determine the starting point of the array in a more efficient way.
+    let paginationMessage: GenericMessage | undefined;
     for (const id of resultIds) {
       const message = await this.get(tenant, id, options);
       if (message) { messages.push(message); }
+      if (pagination?.messageCid && pagination.messageCid === id) {
+        paginationMessage = message;
+      }
+    }
+
+    if (pagination?.messageCid !== undefined && paginationMessage === undefined) {
+      return { messages: [] }; //if paginationMessage is not found, do not return any results
     }
 
-    return messages;
+    const sortedRecords = await MessageStoreLevel.sortMessages(messages, messageSort);
+    return this.paginateRecords(sortedRecords, paginationMessage, pagination);
+  }
+
+  private async paginateRecords(
+    messages: GenericMessage[],
+    paginationMessage?: GenericMessage,
+    pagination: Pagination = { }
+  ): Promise<{ messages: GenericMessage[], paginationMessageCid?: string } > {
+    const { limit } = pagination;
+    if (paginationMessage === undefined && limit === undefined) {
+      return { messages }; // return all without pagination pointer.
+    }
+
+    // this is an optimization, we are passing the pagination message object for an easier lookup
+    // since we know this object exists within the array if passed, we can assume that it will always have a value greater than -1
+    const cursorIndex = paginationMessage ? messages.indexOf(paginationMessage) : undefined;
```

Comment:
Regarding this optimization of finding the cursor index to trim the result set, I would have thought a more performant alternative (if set it large) is to modify the queries themselves by looking up the corresponding sorted property value of the message. Is this considered? I am okay if you agree it's a good idea but want to file a `good-first-issue` and address this in a subsequent PR.

Reply:
Yup! I think after some thought and our discussion about sorting/pagination in general within Level (beyond what it's doing right now) I have some ideas, but they would definitely be suited for their own PR.

I agree that the optimization you mentioned would be a great `good-first-issue` as a first step in the overall pagination/sorting enhancements.

Reply:
Added a TODO to #506 in code.